apps/src/lib/node/ledger/shell/prepare_proposal.rs

```diff
@@ -77,6 +64,231 @@ mod prepare_block {
+                                validator,
+                                err
+                            );
+                        })
+                        .ok()?;
+
+                    // verify signature of the vote extension
+                    let validator_public_key = self
+                        .get_validator_from_address(
+                            &validator_addr,
+                            Some(events_epoch),
+                        )
+                        .map(|(_, validator_public_key)| validator_public_key)
+                        .map_err(|_| {
+                            tracing::error!(
+                                "Could not get public key from Storage for \
+                                 validator {validator_addr}"
+                            );
+                        })
+                        .ok()?;
+
+                    vote_extension
+                        .verify(&validator_public_key)
+                        .map_err(|_| {
+                            tracing::error!(
+                                "Failed to verify the signature of a vote \
+                                 extension issued by {validator_addr}"
+                            );
+                        })
+                        .ok()?;
+
+                    Some((validator_addr, vote_extension))
+                });
+
+            let mut event_observers = BTreeMap::new();
+            let mut signatures = Vec::new();
+
+            let total_voting_power =
+                self.get_total_voting_power(Some(events_epoch)).into();
+            let mut voting_power = 0u64;
+
+            for (validator_addr, vote_extension) in all_vote_extensions {
+                let (validator_voting_power, _) = self
+                    .get_validator_from_address(
+                        &validator_addr,
+                        Some(events_epoch),
+                    )
+                    .expect(
+                        "We already checked that we have a valid Tendermint \
+                         address",
+                    );
+
+                // update voting power
+                voting_power += u64::from(validator_voting_power);
+
+                // register all ethereum events seen by `validator_addr`
+                for ev in vote_extension.data.ethereum_events {
+                    let signers =
+                        event_observers.entry(ev).or_insert_with(HashSet::new);
+
+                    signers.insert(validator_addr.clone());
+                }
+
+                // register the signature of `validator_addr`
+                let addr = validator_addr;
+                let sig = vote_extension.sig;
+
+                signatures.push((sig, addr));
+            }
+
+            let voting_power =
+                FractionalVotingPower::new(voting_power, total_voting_power)
+                    .unwrap();
+
+            if voting_power <= FractionalVotingPower::TWO_THIRDS {
```

Comment:
If this ever happens, we are in the shit.

Reply:
how do we handle this scenario? right now I'm just returning an empty vec of protocol transactions

Reply:
actually this is already covered by the `unreachable!()` I added recently. I'll mark this as solved